### ISSUE TITLE
Completion different from what I select

### ISSUE THREAD

**Issue by dmalyuta** [Sodel-the-Vociferous/helm-company#10]
Sometimes, when I select a given completion, what gets printed in the source code is different. Example:

![wrong_completion_bug](https://cloud.githubusercontent.com/assets/12475811/26794739/c9b3f622-49d7-11e7-8d00-ad6352d6c9d0.gif)

System: Ubuntu 14.04.5, Emacs 25.2.1, `helm-company` master branch.

**Comment by Sodel-the-Vociferous** [comment]
Thank you for the report! Looks like this bug has actually been lurking for a while. Fixed in `master`. Let me know if the fix works for you.

**Comment by dmalyuta** [comment]
This has fixed it for the screencast above, but the bug persists for some proprietary code that I have. In short, I have a C++ structure called `Foobar` and my code looks like this:
```
FooBar bob;
// stuff
myFunction(bob.);
```
I trigger completion after `bob.` and want to select the field `p`. However, selecting it prints `FooBar::` instread :(

**Comment by Sodel-the-Vociferous** [comment]
I keep trying, but I can't reproduce this at all. I've been using this dummy `foo.cpp` file to test this.

    #include <string>
    #include <iostream>
    
    typedef struct FooBar {
        int p;
        int position;
        int positionTwo;
    } FooBar;

    void myFunction(FooBar bob)
    {
    }

    int main()
    {
        FooBar bob;
        myFunction(bob.);
    }

 Is there anything I can change in this test code to make it reproduce your glitch?

**Comment by dmalyuta** [comment]
Here is an example showing two "wrong completion" bugs that persist:

![helm_company_wrong_completion_bug](https://user-images.githubusercontent.com/12475811/26844090-2350396e-4aa7-11e7-8acd-04f8bc779dd9.gif)

- Instead of `#include <header>`, `#include "header"` is selected
- Overloaded function does not complete correctly

I'm not sure why there are these bugs, actually? Selections with the `company-mode` popup menu have always worked and `helm` always selects the correct stuff otherwise - so I'm wondering why `helm-company` has these issues in certain cases.

**Comment by Sodel-the-Vociferous** [comment]
Oh, thanks! I can duplicate the #include completion bug.

The original bug was from a previous author, using the wrong string comparison function somewhere. (They function they used would return a string if your selection was anywhere inside that string. We needed to do an EXACT string match.)

Most of these completions rely on company annotations. (Annotations are the strings that show type information.) Company stores annotations in the original candidate strings as string *properties*. Properties are also how the completion strings get colored. But, when Helm does its completion, it strips all the properties and annotations off of the strings that it returns, which means the annotation and color data gets tossed too.

Basically, Helm throws information away that I need to do the actual completions properly. I just did a bit of a rewrite to more robustly keep that information around. It fixes the #include bug.

I couldn't duplicate the myFunction1() glitch, but I'm hopeful that this rewrite will solve that bug too.

it may even fix #7!

**Comment by Sodel-the-Vociferous** [comment]
Short verion: The #include bug is fixed in`master`. I couldn't reproduce the other bug, so please let me know if it still happens.

**Comment by dmalyuta** [comment]
So far everything works. So I'm closing this, but in case another edge case comes up should I reopen it?

**Comment by Sodel-the-Vociferous** [comment]
Hooray! :) Yes, reopen it if you find another edge case like this one.
